test/retry_test.exs

```diff
@@ -45,44 +48,59 @@ defmodule RetryTest do
       assert elapsed / 1_000 >= 250
     end
 
-    test "retries execution for specified attempts when result is a specified atom" do
-      retry_atom = :not_ok
-
-      {elapsed, _} =
-        :timer.tc(fn ->
-          result =
-            retry with: linear_backoff(50, 1) |> take(5), atoms: [retry_atom] do
-              retry_atom
-            after
-              _ -> :ok
-            else
-              error -> error
-            end
-
-          assert result == retry_atom
-        end)
+    test "retries execution for specified attempts when allowed result is returned" do
+      testcases = [
+        {:not_ok, :all},
+        {:not_ok, [:foo, :all]},
+        {:not_ok, :not_ok},
+        {:not_ok, [:foo, :not_ok]},
+        {{:not_ok, :foo}, [:foo, :not_ok]},
+        {%NotOkay{}, NotOkay},
+        {%NotOkay{}, [Foo, NotOkay]},
+        {:not_ok, fn _ -> true end},
+        {:not_ok, [fn _ -> false end, fn _ -> true end]},
```

Comment:
While I think this is a great feature, I feel like passing a function or a list of functions to `atoms` is semantically weird 🤔. The original idea behind `atoms` is to allowlist things like `:error`, `{:error, foo}`, `:kaput`, etc. Although come to think of it, a more inclusive name for it would be `patterns`. I think we can add `patterns` and deprecate `atoms` if necessary.

Thinking out loud, what if we create a new option called `when`/`when_truthy` that accepts one or more functions instead? I personally think it reads better as well, i.e. "retry when a function that accepts the return value evaluates to true".

Reply:
I agree that `atoms` is no longer the appropriate name, but we obviously don't want to remove `atoms` for backward compatibility reasons so I was hoping we could merge this PR and then maybe in a follow-up add an alias for it and deprecate it.

As far as choosing a new name, I think `when` or `patterns` are both fine. Personally I wouldn't add a third parameter for functions separate from one for atoms (not sure if that's what you were suggesting). If anything, I would go the opposite direction and combing the `rescue_only` functionality in with `when` (or whatever we call it) so you just have a single parameter for deciding when to retry.